### PR TITLE
Support Chainer 2.0.1 and 3.0.0a1

### PR DESCRIPTION
**Pull request by kkk669**
`chainer.links.caffe.CaffeFunction.add_link` has been [deprecated](https://docs.chainer.org/en/latest/reference/generated/chainer.links.caffe.CaffeFunction.html#chainer.links.caffe.CaffeFunction.add_link) since 2.0.0, and `chainer.links.caffe.CaffeFunction` no longer calls `self.add_link` since [this commit](https://github.com/chainer/chainer/pull/2947/files) was merged.

I switched from overriding `add_link` to overriding `__setattr__` so that it can run in Chainer 2.0.1 or later.

This PR is almost the same as [this](https://github.com/Hakuyume/chainer-ssd/pull/10).

### CODE REVIEW

File: examples/ssd/caffe2npz.py [hunk 1]
```diff
@@ -59,17 +61,18 @@ def add_link(self, name, link):
                 data = data.reshape((-1, 4) + data.shape[1:])
                 data[:, [1, 0, 3, 2]] = data.copy()
             print('{:s} -> {:s} (xy -> yx)'.format(name, new_name))
-        else:
+        elif match:
```
Comment (Hakuyume) [inline examples/ssd/caffe2npz.py]: Could you remove `match`? It is better to show all layers even if their names are not changed. And this modification is not related to supporting v2 and v3.

Comment (kkk669) [inline examples/ssd/caffe2npz.py]: Yes, but the reason why I added `match` is because it shows all attributes even if they are not layers (e.g. [self.verbose](https://github.com/Hakuyume/chainer-ssd/blob/master/lib/caffe.py#L45)).

Comment (Hakuyume) [inline examples/ssd/caffe2npz.py]: I see. How about checking `self.within_init_scope` (https://github.com/chainer/chainer/blob/master/chainer/link.py#L153)? This property becomes `True` if the attribute is a layer (a child link).

Comment (Hakuyume) [inline examples/ssd/caffe2npz.py]: Using `within_init_scope`, the code will be more simple.
```
def __setattr__(self, name, link):
    if self.within_init_scope:
        # link is a layer (a child link)
    else:
        # link is a regular attribute
```

Comment (kkk669) [inline examples/ssd/caffe2npz.py]: I didn't know it. Thank you. I'll remove match and use it.

**Comment by Hakuyume** [comment]
@kkk669 Thank you for updating this PR. It looks good to me. Travis CI is not working well now (This is not only for this PR but for all branches). After it is fixed, I will merge  this PR.

**Comment by kkk669** [comment]
@Hakuyume I understand. My final commit is `81d1147` and I'm not planning to push future commits. Thank you for all your advices.

File: examples/ssd/caffe2npz.py [hunk 2]
```diff
@@ -47,7 +48,7 @@ def __init__(self, model_path):
         super(SSDCaffeFunction, self).__init__(model_path)
 
     def __setattr__(self, name, link):
-        if self.within_init_scope and name != '_within_init_scoep':
+        if self.within_init_scope and instanceof(link, Link):
```
Comment (Hakuyume) [inline examples/ssd/caffe2npz.py]: `instanceof` -> `isinstance`
https://github.com/chainer/chainer/blob/master/chainer/link.py#L655

Comment (Hakuyume) [inline examples/ssd/caffe2npz.py]: Could you rename `link` -> `value`? `Chain.__setattr__` uses this name. (https://github.com/chainer/chainer/blob/master/chainer/link.py#L655)

Comment (kkk669) [inline examples/ssd/caffe2npz.py]: I'm sorry for my typo.

Comment (kkk669) [inline examples/ssd/caffe2npz.py]: I've done.

Comment (Hakuyume) [inline examples/ssd/caffe2npz.py]: Thank you!
